src/main/java/cuchaz/enigma/command/MappingCommandsUtil.java

```diff
@@ -0,0 +1,134 @@
+package cuchaz.enigma.command;
+
+import cuchaz.enigma.ProgressListener;
+import cuchaz.enigma.throwables.MappingParseException;
+import cuchaz.enigma.translation.mapping.EntryMapping;
+import cuchaz.enigma.translation.mapping.serde.EnigmaMappingsReader;
+import cuchaz.enigma.translation.mapping.serde.EnigmaMappingsWriter;
+import cuchaz.enigma.translation.mapping.serde.TinyMappingsReader;
+import cuchaz.enigma.translation.mapping.serde.TinyMappingsWriter;
+import cuchaz.enigma.translation.mapping.tree.EntryTree;
+import cuchaz.enigma.translation.mapping.tree.EntryTreeNode;
+import cuchaz.enigma.translation.mapping.tree.HashEntryTree;
+import cuchaz.enigma.translation.representation.entry.ClassEntry;
+import cuchaz.enigma.translation.representation.entry.Entry;
+import cuchaz.enigma.translation.representation.entry.FieldEntry;
+import cuchaz.enigma.translation.representation.entry.MethodEntry;
+
+import java.io.IOException;
+import java.nio.file.Path;
+import java.util.*;
+
+public final class MappingCommandsUtil {
+    private MappingCommandsUtil() {}
+
+    @SuppressWarnings("unchecked")
+    public static EntryTree<EntryMapping> invert(EntryTree<EntryMapping> mappings) {
+        EntryTree<EntryMapping> result = new HashEntryTree<>();
+
+        Map<Entry<?>, Entry<?>> rightEntries = new HashMap<>();
+        for (EntryTreeNode<EntryMapping> node : mappings) {
+            Entry<?> leftEntry = node.getEntry();
+            EntryMapping leftMapping = node.getValue();
+
+            if (!(leftEntry instanceof ClassEntry || leftEntry instanceof MethodEntry || leftEntry instanceof FieldEntry)) {
+                result.insert(((Entry<Entry<?>>) leftEntry).withParent(rightEntries.get(leftEntry.getParent())), leftMapping);
+                continue;
+            }
+
+            Entry<?> rightEntry = ((Entry<Entry<?>>) leftEntry)
+                    .withParent(rightEntries.get(leftEntry.getParent()))
+                    .withName(getInnerName(leftMapping == null ? leftEntry.getName() : leftMapping.getTargetName()));
+            rightEntries.put(leftEntry, rightEntry);
+
+            result.insert(rightEntry, leftMapping == null ? null : new EntryMapping(leftEntry.getName())); // TODO: leftMapping.withName once javadoc PR is merged
+        }
+
+        return result;
```

Comment:
Just curious, where did you handle the descriptor remapping?

Reply:
The EntryTree handles it automatically.